IPNI.md

```diff
@@ -310,6 +308,17 @@ Specified protocols are expected to be ordered in increasing order.
 
 If the `Metadata` field is not specified, the advertisement is treated as address update only.
 
+#### SeqNum
+
+The sequence number field is optional only for backwards compatibility. It should be included as soon as possible by both new and existing index providers to enable indexer functionality that uses it. Omitting the `SeqNum` will result in degraded indexer functionality relating to distance calculation, reporting, indexer synchronization. Future indexer versions may choose to only accept new advertisements that have a sequence number.
+
+A `SeqNum` value higher then the maximum of 2<sup>53</sup>-1 (9007199254740992) values will not be recognized and instead seen as 0. It is not expected for any chain to reach this
+  sequence as that would mean it has over nine quintillion advertisements. If this were reached, then `SeqNum` should be reset to 0, which may result in an inaccurate distance calculations between advertisements spanning the reset.
+
+If `SeqNum` is present then it is included in the data that the advertisement signature is calculated over.
+
+Indexers are not required to validate that `SeqNum` is monotonically increasing, but an incorrect sequence number may result in incorrect indexing information or may be ignored resulting in some indexing features being unavailable for that chain's provider. If an indexer depends on the sequence number for more than display and reporting purposes, it can choose not index advertisement chains that do not have a monotonically increasing `SeqNum`.
```

Comment:
I think this is too forgiving; For sequence number to be useful it should provide guarantees to the observer.

Reply:
We should decide what is the correct way to handle an incorrect sequence number. Other than the above the above policy of tolerance, I see three ways to deal with bad sequence:
1. Ignore all ads until an ad has the next expected sequence is put on the chain. For example:
  1 2 9 4 5 6 3 4 5 6...  Here 9 4 5 6 are ignored and indexing is resumed at 4
2. Ignore all ads that do not match the expected sequence. Example:
   Expect: 1 2 3 4 5 6 7 8 ...
   Have:    1 2 9 4 5 6 3 4 ...
   Here 9 is ignored, then 3 4 and all that follow are ignored until the sequence lines up with expected.
3. Ignore all ads until sequence number is reset to 0.
4. Stop indexing chain and require a new chain be created.

The first two can prevent different indexers from reaching consistency, since depending on where they start indexing, the errors may look different. Also, either of these strategies can cause the remainder of the chain to be ignored until the _appropriate_ correction is put into an advertisement.

Number 3 also ignores the chain until a specific correction (reset to 0) is put into an advertisement. This also makes sequence completely optional by allowing a zero value that resets sequence with each ad.

Number 4 seems unnecessarily harsh and wasteful, and may not be something that providers will be willing to even do.

So, that leaves us with the original strategy of tolerance. A wrong sequence may cause some unreliable distance computations when computing the distance of ads that span one or more bad sequence numbers. However enough ads with monotonically increasing sequence numbers are added to the chain, these errors will disappear as they will be sufficiently in the past that all but very far behind indexers will even see them. Tolerance leads to a self-correcting system that will allow different indexers to reach consistency.

Reply:
Add to the list Option 5:
* No. 4 on providers side in that once they decide to use sequence numbers they should stick to it and increment it correctly. 
* Do height calculation at indexers as starting point.

>Tolerance leads to a self-correcting system

...with the right incentive. I am not sure just now what that is.

Another question in my head is:
* how different would incorrect sequence number be compared to incorrect signature? Providers don't even think about the signature. They just follow the protocol. What specific failure scenario we are thinking of that makes sequence number require tolerance.

Maybe it's worth fleshing those failure scenarios out.